src/annotator/util/buckets.js

```diff
@@ -0,0 +1,248 @@
+ */
+
+/**
+ * An object containing information about anchor highlight positions
+ *
+ * @typedef PositionPoints
+ * @prop {Anchor[]} above - Anchors that are offscreen above
+ * @prop {Anchor[]} below - Anchors that are offscreen below
+ * @prop {PositionPoint[]} points - PositionPoints for on-screen anchor
+ *   highlights. Each highlight box has 2 PositionPoints (one for top edge
+ *   and one for bottom edge).
+ */
+
+/**
+ * @typedef BucketInfo
+ * @prop {Array<Anchor[]>} buckets
+ * @prop {number[]} index - Array of (pixel) positions of each bucket
+ */
+
+// FIXME: Temporary duplication of size constants between here and BucketBar
+const BUCKET_SIZE = 16; // Regular bucket size
+const BUCKET_NAV_SIZE = BUCKET_SIZE + 6; // Bucket plus arrow (up/down)
+const BUCKET_TOP_THRESHOLD = 115 + BUCKET_NAV_SIZE; // Toolbar
+
+// TODO!! This is an option in the plugin right now
+const BUCKET_GAP_SIZE = 60;
+
+/**
+ * Find the closest valid anchor in `anchors` that is offscreen in the direction
+ * indicated.
+ *
+ * @param {Anchor[]} anchors
+ * @param {'up'|'down'} direction
+ * @return {Anchor|null} - The closest anchor or `null` if no valid anchor found
+ */
+export function findClosestOffscreenAnchor(anchors, direction) {
+  let closestAnchor = null;
+  let closestTop = 0;
+
+  for (let anchor of anchors) {
+    if (!anchor.highlights?.length) {
+      continue;
+    }
+
+    const top = getBoundingClientRect(anchor.highlights).top;
+
+    // Verify that the anchor is offscreen in the direction we're headed
+    if (direction === 'up' && top >= BUCKET_TOP_THRESHOLD) {
+      // We're headed up but the anchor is already below the
+      // visible top of the bucket bar: it's not our guy
+      continue;
+    } else if (
+      direction === 'down' &&
+      top <= window.innerHeight - BUCKET_NAV_SIZE
+    ) {
+      // We're headed down but this anchor is already above
+      // the usable bottom of the screen: it's not our guy
+      continue;
+    }
+
+    if (
+      !closestAnchor ||
+      (direction === 'up' && top > closestTop) ||
+      (direction === 'down' && top < closestTop)
+    ) {
+      // This anchor is either:
+      // - The first anchor we've encountered off-screen in the direction
+      //   we're headed, or
+      // - Closer to the screen than the previous `closestAnchor`
+      closestAnchor = anchor;
+      closestTop = top;
+    }
+  }
+
+  return closestAnchor;
+}
+
+/**
+ * Construct an Array of points representing the positional tops and bottoms
+ * of current anchor highlights. Each anchor whose highlight(s)' bounding
+ * box is onscreen will result in two entries in the `points` Array: one
+ * for the top of the highlight box and one for the bottom
+ *
+ * @param {Anchor[]} anchors
+ * @return {PositionPoints}
+ */
+export function constructPositionPoints(anchors) {
+  const aboveScreenAnchors = new Set();
+  const belowScreenAnchors = new Set();
+  const points = /** @type {PositionPoint[]} */ (new Array());
```

Comment:
```suggestion
  const points = /** @type {PositionPoint[]} */ ([]);
```

Small stylistic suggestion.

Reply:
And a good one. I have a reason for why this happened (I'd typically use `[]` when initializing an Array myself), but it's boring.

Reply:
These have been changed.